kvcoord: condense read spans when they exceed the memory limit

andreimatei opened the pull request:
Before this patch, once a transaction exceeds the
kv.transaction.max_refresh_spans_bytes limit, it stopped tracking reads
and it didn't attempt to refresh any more when pushed.
This patch make the span refresher condense the spans when it runs out
of memory instead. So we'll get bigger spans and potentially false
conflicts, but at least we have a chance at refreshing. In particular,
it'll succeed if there's no writes anywhere.

The condensing is performed using the condensableSpanSet, like we do in
the pipeliner interceptor for the tracking of write intents. Internally,
that guy condenses spans in ranges with lots of reads.

We've seen people run into kv.transaction.max_refresh_spans_bytes in the
past, so this should help many uses cases. But in particular I've
written this patch because, without it, I'm scared about the effects of
20.1's reduction in the closed timestamp target duration to 3s from a
previous 30s. Every transaction writing something after having run for
longer than that will get pushed, so being able to refresh is getting
more important.

Fixes #46095

Release note (general change): Transactions reading a lot of data behave
better when exceeding the memory limit set by
kv.transaction.max_refresh_spans_bytes. Such transactions now attempt to
resolve the conflicts they run into instead of being forced to always
retry. Increasing kv.transaction.max_refresh_spans_bytes should no
longer be necessary for most workloads.

Release justification: fix for new "functionality" - the reduction in
the closed timestamp target duration.

andreimatei committed:
kv: save allocation in read span copying

This patch removes a scary comment and some code that has become
unnecessary a while ago. It had to do with protecting some state that's
no longer shared.

Release note: None

andreimatei committed:
kv: elevate message to a warning

Trouble with range iteration can cause a transaction to fail to collapse
its write footprint. This is unexpected enough, and also not handled
well, so let's Warningf. Particularly since we're about to start using
this code to collapse the read footprint too.

Release note: None

andreimatei committed:
pkg: provide ctx to an interceptor function

It's needed by the next commit.

Release note: None

andreimatei committed:
kvcoord: change the error type returned by RangeIter

The range iterator used to return pErr. This patch changes it to regular
error. The RangeIter is used more and more outside of the DistSender.
Almost all the callers already wanted regular errors. There was no
benefit to returning pErr from it. More over, the next commit want to
abstract an interface from it, and so pErr really doesn't belong.

Release note: None

andreimatei committed:
kvcoord: fix a size tracking bug in condensableSpanSet

The span set was misinterpreting the return of roachpb.MergeSpans().

Release note: None

andreimatei committed:
kvcoord: condense read spans when they exceed the memory limit

Before this patch, once a transaction exceeds the
kv.transaction.max_refresh_spans_bytes limit, it stopped tracking reads
and it didn't attempt to refresh any more when pushed.
This patch make the span refresher condense the spans when it runs out
of memory instead. So we'll get bigger spans and potentially false
conflicts, but at least we have a chance at refreshing. In particular,
it'll succeed if there's no writes anywhere.

The condensing is performed using the condensableSpanSet, like we do in
the pipeliner interceptor for the tracking of write intents. Internally,
that guy condenses spans in ranges with lots of reads.

We've seen people run into kv.transaction.max_refresh_spans_bytes in the
past, so this should help many uses cases. But in particular I've
written this patch because, without it, I'm scared about the effects of
20.1's reduction in the closed timestamp target duration to 3s from a
previous 30s. Every transaction writing something after having run for
longer than that will get pushed, so being able to refresh is getting
more important.

Fixes #46095

Release note (general change): Transactions reading a lot of data behave
better when exceeding the memory limit set by
kv.transaction.max_refresh_spans_bytes. Such transactions now attempt to
resolve the conflicts they run into instead of being forced to always
retry. Increasing kv.transaction.max_refresh_spans_bytes should no
longer be necessary for most workloads.

Release justification: fix for new "functionality" - the reduction in
the closed timestamp target duration.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46275)
<!-- Reviewable:end -->

nvb reviewed (commented):
For the release note:
- I think you need to give it a category
- s/Transsactions/Transactions/

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 7 of 7 files at r3, 5 of 5 files at r4, 2 of 2 files at r5, 11 of 11 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-:-M36jVEfDEzCmECX4HtF:bamsrep)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 1678 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36lZMrB7tc_PuHAZ3c:-M36lZMrB7tc_PuHAZ3d:b-fzl28u) ([raw file](https://github.com/cockroachdb/cockroach/blob/733563a521a9323f77411a609af1ea019073c6c9/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L1678)):*
> ```Go
> 		},
> 		{
> 			// If we've exhausted the limit for tracking refresh spans but we
> ```

We should try to keep these tests somehow. We don't want to lose testing coverage, even if we expect the number of cases where the refresh spans become invalid to decrease.
___
*[pkg/kv/kvclient/kvcoord/range_iter.go, line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36kQxC7dF_1lCmCOxR:-M36kQxC7dF_1lCmCOxS:b-el9cg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e63d377e1fb6eeedc41c94cf56dcf4d39863104/pkg/kv/kvclient/kvcoord/range_iter.go#L206)):*
> ```Go
> 
> 	// Check for an early exit from the retry loop.
> 	if pErr := ri.ds.deduceRetryEarlyExitError(ctx); pErr != nil {
> ```

If you change `deduceRetryEarlyExitError` to return an `error`, which looks possible without any casts, then you won't need any `GoError()` calls, which are always a bit suspicious.
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender.go, line 220 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36mDEY3uEugmAYwl7-:-M36mDEZCVM9GCDiReez:bckk1gj) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L220)):*
> ```Go
> 	// TxnCoordSender's wrapped sender. First, each of the interceptor objects
> 	// is initialized.
> 	var riGen rangeIteratorFactory
> ```

Why isn't this in `initCommonInterceptors`?
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 890 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36kx0pEP_AsM_jQ-LH:-M36kx0pEP_AsM_jQ-LI:b-3l83tg) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a99d5c80d2fa17c66c3a9e8c91c0fb7d5a09b7/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L890)):*
> ```Go
> // The method has the side effect of sorting the stable write set.
> func (s *condensableSpanSet) mergeAndSort() (distinct bool) {
> 	oldLen := len(s.s)
> ```

Is it worth writing a test for this fix? It shouldn't be too much work given how close we're are to this code in `txn_interceptor_pipeliner_test.go`.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 181 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36mPtA8ih_gnlUVZH_:-M36mPtA8ih_gnlUVZHa:b-yrwb5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L181)):*
> ```Go
> type txnPipeliner struct {
> 	st       *cluster.Settings
> 	riGen    rangeIteratorFactory //used to condense lock spans, if provided
> ```

"// used"
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 207 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36meTU-KDhY2f1Izm-:-M36meTU-KDhY2f1Izm0:b-3zd5mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L207)):*
> ```Go
> }
> 
> // condensableSpanSetRangeIterator describes the interface of RangeIterator
> ```

Does this belong in `txn_interceptor_pipeliner.go`? If the rangeIteratorFactory and `condensableSpanSet` are used in multiple interceptors then I'd put them either a) in their own file or b) in `txn_coord_sender.go`.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 220 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36mx6N1QZRtmoxpRhG:-M36mx6O85NuSPe5R11G:b-mjv5y4) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L220)):*
> ```Go
> // lazily. It's used to avoid allocating an iterator when it's not needed. The
> // factory can be configured either with a callback, used for mocking in tests,
> // or with a DistSender. Can also not be configured with anything, in which
> ```

This sentence is missing an ending.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 920 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36nPXPEdLxk1NXcp3F:-M36nPXQE4yzeRZht__J:bf5qh6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L920)):*
> ```Go
> }
> 
> func (s *condensableSpanSet) insertMultiple(spans []roachpb.Span) {
> ```

Might as well add a comment here like we have for `insert`.

Also, any reason not to use a variadic arg?
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 951 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36njo7ASVva1evHY47:-M36njo7ASVva1evHY48:b4099e6) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L951)):*
> ```Go
> // spans within the same range boundaries are also condensed.
> //
> // Returns true if condensing was necessary.
> ```

Right now it returns true if condensing was performed, not if it was necessary. Should we return true from the "failed to condense lock spans" case or update the comment?
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 1048 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36nylc8bTh2-OukvP5:-M36nyld6Mgh8wA0_T-K:bwf76ea) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L1048)):*
> ```Go
> func (s *condensableSpanSet) clear() {
> 	s.s = nil
> 	s.bytes = 0
> ```

`s.condensed = false`? 

Or just `*s = condensableSpanSet{}` to avoid this kind of bug.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner_test.go, line 563 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36oBVq-DAYbR-2oUDo:-M36oBVq-DAYbR-2oUDp:bf2vo4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner_test.go#L563)):*
> ```Go
> 	tp, mockSender := makeMockTxnPipeliner()
> 
> 	// Disable maxInFlightSize, maxBatchSize and max_intents_bytes limits.
> ```

", and"

Also, could you make this either all refer to the setting names or all refer to the variable names?
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36o_Az3lK4mU8ExJ4r:-M36o_Az3lK4mU8ExJ4s:b-z0aa71) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a99d5c80d2fa17c66c3a9e8c91c0fb7d5a09b7/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L109)):*
> ```Go
> 	wrapped lockedSender
> 
> 	// refreshSpans contains key spans which were read during the  transaction. In
> ```

Why remove this comment?
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 110 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36od_83Hb-PZVB0t9V:-M36od_83Hb-PZVB0t9W:b-nkcg8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L110)):*
> ```Go
> 
> 	refreshFootprint condensableSpanSet
> 	riGen            rangeIteratorFactory
> ```

Move this above `wrapped`. It's a dependency of the struct.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 115 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36or1p7HRjl4nQ_o2V:-M36or1p7HRjl4nQ_o2W:b-uj8pqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L115)):*
> ```
> 	// Although maybe we should still be using it when
> 	// condensing the spans doesn't reduce them as much as we want.
> ```

I'm surprised to find that we aren't. Why is that?
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36pUeh4kEBONypFap-:-M36pUeh4kEBONypFap0:b973z6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L338)):*
> ```Go
> 	refreshSpanBa := roachpb.BatchRequest{}
> 	refreshSpanBa.Txn = refreshTxn
> 	addRefreshes := func(refreshes condensableSpanSet) {
> ```

nit: it feels more natural for this to be a pointer.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go, line 772 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36poczBvTBOECdSLb6:-M36poczBvTBOECdSLb7:b-5acia) ([raw file](https://github.com/cockroachdb/cockroach/blob/733563a521a9323f77411a609af1ea019073c6c9/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go#L772)):*
> ```Go
> 	require.Equal(t, hlc.Timestamp{}, tsr.refreshedTimestamp)
> 
> 	// Send a batch above the limit.
> ```

It's unfortunate that we're losing this test coverage as well.
___
*[pkg/kv/kvclient/kvcoord/txn_metrics.go, line 76 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36q9VM-wKGHzY6Owlc:-M36q9VM-wKGHzY6Owld:b-d9wekj) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_metrics.go#L76)):*
> ```Go
> 	}
> 	metaRefreshSpanBytesExceeded = metric.Metadata{
> 		Name: "txn.refreshspanbytesexceeded",
> ```

Does this metric make sense anymore? Now that exceeding the byte limit does not imply auto-failing, should we break it into:
1. a metric about the # txns that exceeded byte limit
2. a metric about the # txns that failed to refresh
___
*[pkg/roachpb/merge_spans.go, line 40 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36kmFB7KQOHHfYkRRl:-M36kmFB7KQOHHfYkRRm:b-trms7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a99d5c80d2fa17c66c3a9e8c91c0fb7d5a09b7/pkg/roachpb/merge_spans.go#L40)):*
> ```Go
> // The input spans are not safe for re-use.
> //
> // Note that even if it returns true, adjacent spans might have been merged
> ```

nit: move this right after the sentence about the return value.


<!-- Sent from Reviewable.io -->

andreimatei committed:
kvcoord: move condensableSpanSet to its own file

... in anticipation of it being used by a second interceptor.

Release note: None

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 1678 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36lZMrB7tc_PuHAZ3c:-M38qXbX6GFHX76QqJoe:bvh6q7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/733563a521a9323f77411a609af1ea019073c6c9/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L1678)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We should try to keep these tests somehow. We don't want to lose testing coverage, even if we expect the number of cases where the refresh spans become invalid to decrease.
</blockquote></details>

ok, I've added a knob to disable condensing and added back these tests
___
*[pkg/kv/kvclient/kvcoord/range_iter.go, line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36kQxC7dF_1lCmCOxR:-M37i1Qn69Y-w0N2aJHV:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e63d377e1fb6eeedc41c94cf56dcf4d39863104/pkg/kv/kvclient/kvcoord/range_iter.go#L206)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If you change `deduceRetryEarlyExitError` to return an `error`, which looks possible without any casts, then you won't need any `GoError()` calls, which are always a bit suspicious.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender.go, line 220 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36mDEY3uEugmAYwl7-:-M37kBfy2my8vneoKmQ2:b-7mmp82) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L220)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why isn't this in `initCommonInterceptors`?
</blockquote></details>

moved
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 890 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36kx0pEP_AsM_jQ-LH:-M38NP6T3ovQ596iTz0P:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a99d5c80d2fa17c66c3a9e8c91c0fb7d5a09b7/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L890)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is it worth writing a test for this fix? It shouldn't be too much work given how close we're are to this code in `txn_interceptor_pipeliner_test.go`.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 181 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36mPtA8ih_gnlUVZH_:-M37kOO3D4_5_cg-8LmS:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L181)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"// used"
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 207 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36meTU-KDhY2f1Izm-:-M38MAEw0HJEGePa4Kd2:b-d3sl0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L207)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this belong in `txn_interceptor_pipeliner.go`? If the rangeIteratorFactory and `condensableSpanSet` are used in multiple interceptors then I'd put them either a) in their own file or b) in `txn_coord_sender.go`.
</blockquote></details>

moved to a new file in a new commit
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 220 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36mx6N1QZRtmoxpRhG:-M37lZQGBU8VINPVHF00:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L220)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This sentence is missing an ending.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 920 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36nPXPEdLxk1NXcp3F:-M38NMml4_d0Thusz1k_:b-hfiwgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L920)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Might as well add a comment here like we have for `insert`.

Also, any reason not to use a variadic arg?
</blockquote></details>

switch to variadic arg
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 951 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36njo7ASVva1evHY47:-M38NWX7CTR6OR0PInzB:blwyhcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L951)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Right now it returns true if condensing was performed, not if it was necessary. Should we return true from the "failed to condense lock spans" case or update the comment?
</blockquote></details>

updated comment
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 1048 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36nylc8bTh2-OukvP5:-M38NzHNEiJpgOBIjZ2a:bd9adub) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L1048)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s.condensed = false`? 

Or just `*s = condensableSpanSet{}` to avoid this kind of bug.
</blockquote></details>

done, good catch
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner_test.go, line 563 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36oBVq-DAYbR-2oUDo:-M38ODsyDB1vplDaWJEA:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner_test.go#L563)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

", and"

Also, could you make this either all refer to the setting names or all refer to the variable names?
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36o_Az3lK4mU8ExJ4r:-M37jGZu0nP0BqK7S2ej:b5j00gg) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a99d5c80d2fa17c66c3a9e8c91c0fb7d5a09b7/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L109)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why remove this comment?
</blockquote></details>

reinstated
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 110 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36od_83Hb-PZVB0t9V:-M37jNHh1Yg5YLseaUnc:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L110)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Move this above `wrapped`. It's a dependency of the struct.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 115 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36or1p7HRjl4nQ_o2V:-M38RgFA5G-1tiS1txbJ:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L115)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> 	// Although maybe we should still be using it when
> 	// condensing the spans doesn't reduce them as much as we want.
> ```

I'm surprised to find that we aren't. Why is that?
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36pUeh4kEBONypFap-:-M37jV-rErVf0mu6aMoe:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L338)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: it feels more natural for this to be a pointer.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go, line 772 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36poczBvTBOECdSLb6:-M3CXVt08ddR67RsJURw:bb4ugsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/733563a521a9323f77411a609af1ea019073c6c9/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go#L772)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's unfortunate that we're losing this test coverage as well.
</blockquote></details>

put it back
___
*[pkg/kv/kvclient/kvcoord/txn_metrics.go, line 76 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36q9VM-wKGHzY6Owlc:-M3Ccwt6AGp_JWp3DV3Z:b-2x4j6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_metrics.go#L76)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this metric make sense anymore? Now that exceeding the byte limit does not imply auto-failing, should we break it into:
1. a metric about the # txns that exceeded byte limit
2. a metric about the # txns that failed to refresh
</blockquote></details>

see now
___
*[pkg/roachpb/merge_spans.go, line 40 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36kmFB7KQOHHfYkRRl:-M38IOiv2iVkcfpNB8Gc:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a99d5c80d2fa17c66c3a9e8c91c0fb7d5a09b7/pkg/roachpb/merge_spans.go#L40)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: move this right after the sentence about the return value.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
This is close to an <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, but there seems to be stuff missing. Do you have a few changes locally that weren't pushed?

> Reviewed 19 of 19 files at r7, 1 of 1 files at r8, 7 of 7 files at r9, 5 of 5 files at r10, 3 of 3 files at r11, 4 of 4 files at r12, 14 of 14 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-:-M3g7gqd3JKxCWnSw_FP:boii3h4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 1760 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M3g6Qq63SFrmWrix7gN:-M3g6Qq63SFrmWrix7gO:b-5eenww) ([raw file](https://github.com/cockroachdb/cockroach/blob/34dd44eae73fb7d4190646e1ed88fbe56c6bfec0/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L1760)):*
> ```Go
> 			// error. This is the case even if the final batch itself causes a
> 			// refresh.
> 			name:                       "forwarded timestamp with too many refreshes in batch commit triggering refresh",
> ```

What's up with this spacing? Does this need a `gofmt`?
___
*[pkg/kv/kvclient/kvcoord/range_iter.go, line 203 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M3g6h0cBs2T5dBgqyzn:-M3g6h0cBs2T5dBgqyzo:bi1n63w) ([raw file](https://github.com/cockroachdb/cockroach/blob/34dd44eae73fb7d4190646e1ed88fbe56c6bfec0/pkg/kv/kvclient/kvcoord/range_iter.go#L203)):*
> ```Go
> 
> 	// Check for an early exit from the retry loop.
> 	if err := ri.ds.deduceRetryEarlyExitError(ctx); err != nil {
> ```

nit: do the intermediate commits all compile?
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 181 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36mPtA8ih_gnlUVZH_:-M3g6rsoEyLjGeQxvte8:bfhcj3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L181)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

done
</blockquote></details>

Is this done?
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36o_Az3lK4mU8ExJ4r:-M3g75tgEgDBC2wHJJ93:balydiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a99d5c80d2fa17c66c3a9e8c91c0fb7d5a09b7/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L109)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

reinstated
</blockquote></details>

Was it?
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36pUeh4kEBONypFap-:-M3g7IPXDx_ejnDPsQ8N:bulygo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L338)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

done
</blockquote></details>

Is it?
___
*[pkg/kv/kvclient/kvcoord/txn_metrics.go, line 75 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M3g7RaPCZdZgSCk-vNn:-M3g7RaPCZdZgSCk-vNo:b-m9snjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/34dd44eae73fb7d4190646e1ed88fbe56c6bfec0/pkg/kv/kvclient/kvcoord/txn_metrics.go#L75)):*
> ```Go
> 	metaRefreshSuccess = metric.Metadata{
> 		Name:        "txn.refresh.success",
> 		Help:        "Number of successful refreshes.",
> ```

Do these "Help" texts usually have punctuation? It doesn't look like it.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46275)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 1760 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M3g6Qq63SFrmWrix7gN:-M3hqcNU8i4tVpWcvny6:b-zf9tx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/34dd44eae73fb7d4190646e1ed88fbe56c6bfec0/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L1760)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's up with this spacing? Does this need a `gofmt`?
</blockquote></details>

well, the formatter aligns it with the next field - which has a long name. I've done something.
___
*[pkg/kv/kvclient/kvcoord/range_iter.go, line 203 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M3g6h0cBs2T5dBgqyzn:-M3hyRUdEmycNd7gzey0:b-8gs9gv) ([raw file](https://github.com/cockroachdb/cockroach/blob/34dd44eae73fb7d4190646e1ed88fbe56c6bfec0/pkg/kv/kvclient/kvcoord/range_iter.go#L203)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: do the intermediate commits all compile?
</blockquote></details>

they do. It's not many people that I would have checked for.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go, line 181 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36mPtA8ih_gnlUVZH_:-M3hr-yc3Hjhoa_F6AiF:b-i1jon6) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L181)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this done?
</blockquote></details>

now it is... I had stashed some stuff and never put it back...
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36o_Az3lK4mU8ExJ4r:-M3ht72tCuJ8AIEJAoCq:b-p9pl23) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a99d5c80d2fa17c66c3a9e8c91c0fb7d5a09b7/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L109)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Was it?
</blockquote></details>

is now
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M36pUeh4kEBONypFap-:-M3hrETO10q16LGpoiDl:bcd155q) ([raw file](https://github.com/cockroachdb/cockroach/blob/571f8f9beead3b42d239dbf7a0b2e8bf6bef8a26/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L338)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is it?
</blockquote></details>

now it is...
___
*[pkg/kv/kvclient/kvcoord/txn_metrics.go, line 75 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M3g7RaPCZdZgSCk-vNn:-M3hrKhF3TmdMiLf61l_:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/34dd44eae73fb7d4190646e1ed88fbe56c6bfec0/pkg/kv/kvclient/kvcoord/txn_metrics.go#L75)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do these "Help" texts usually have punctuation? It doesn't look like it.
</blockquote></details>

fixed


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 4 of 7 files at r16, 3 of 5 files at r17, 1 of 3 files at r18, 1 of 4 files at r19, 14 of 14 files at r20.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-:-M3ifsbV7Jh3m9cKvr57:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/kv/kvclient/kvcoord/txn_metrics.go, line 75 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M3g7RaPCZdZgSCk-vNn:-M3ifmWL4WLw0AikbU3M:budxoah) ([raw file](https://github.com/cockroachdb/cockroach/blob/34dd44eae73fb7d4190646e1ed88fbe56c6bfec0/pkg/kv/kvclient/kvcoord/txn_metrics.go#L75)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

fixed
</blockquote></details>

Same issue below.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-:-M3lAUl689kGLuMoxAyB:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/kvclient/kvcoord/txn_metrics.go, line 75 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46275#-M3g7RaPCZdZgSCk-vNn:-M3lARYF-8JwpxmNEQTR:b-qssebl) ([raw file](https://github.com/cockroachdb/cockroach/blob/34dd44eae73fb7d4190646e1ed88fbe56c6bfec0/pkg/kv/kvclient/kvcoord/txn_metrics.go#L75)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same issue below.
</blockquote></details>

fixed.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reminder to backport this to `release-20.1` before checking off the associated release blocker.
